src/lambda/cache.rs

```diff
@@ -14,7 +15,7 @@ pub trait CacheKey<Ctx> {
 #[derive(Clone, Debug)]
 pub struct Cache {
     pub max_age: NonZeroU64,
-    pub expr: IO,
+    pub expr: Box<Expression>,
```

Comment:
We wanted to make sure that cache can only be applied on IOs. Why have we changed this guarantee?

Reply:
because in `src/lambda/modify.rs` we can pass functions that works only on expression and to call it recursively we need to pass the expression not io to support typings.

Reply:
If I understand you, are you saying we need to recursively wrap the expression with Cache expression?

I think that might not be required. We could insert the cache operator at the first IO express, and cache just that. Any IO that's happening internally shouldn't need any caching because the top IO expression just resolve from cache.

It's possible that I misunderstand your point here. Feel free to expand more with an example if you think I am wrong.

Reply:
no, it's not about that, it's more about type and traits we have.

1. In app_ctx we want to traverse all the field resolvers and update io operations.
2. The code https://github.com/tailcallhq/tailcall/blob/51d8b7aff838f0f4c362d4ee9e39492ae1f51fdb/src/app_context.rs#L37-L39 use `Expression::modify` that accepts callback `Expression -> Expression`
3. In order to update only specific fields we have that modify logic that recursively traverse expression and calls the callback https://github.com/tailcallhq/tailcall/blob/a12599f0fe916835bb04e46c28339f956d06b90b/src/lambda/modify.rs#L41-L43 Note here we return new IO expr but we actually should return Expression::Cache
4. But the `modifier` can actually return any expression (according to types) not only IO and we need somehow to unwrap returned IO from Expression.

Because the modifier returns wrapper `Expression` but cache stores only IO I think the mistake was done in the first place i.e. types are misleading. 
We could solve by having something specific to handle io operations or have in-place updates how I was doing in the https://github.com/tailcallhq/tailcall/pull/1032/files